pyluna-pathology/luna/pathology/cli/slide_etl.py

```diff
@@ -0,0 +1,188 @@
+    """ CLI tool method
+
+    Args:
+        input_slide_folder (str): path to parent directory containing slide images
+        project_name (str): project name underwhich the slides should reside
+        comment (str): comment and description of dataset
+        num_cores (int): number of cores (interfaced to dask)
+        debug_limit (int): cap number fo slides to process below this number (for debugging, testing, dry_run is automatically enabled)
+        dry_run (bool): don't actually generate or copy any data on disk, print outputs
+        output_dir (str): path to output table
+
+
+    Returns:
+        dict: metadata about function call
+    """
+    slide_paths = []
+
+    for path, _, files in os.walk(input_slide_folder):
+        for file in files:
+            file = os.path.join(path, file)
+
+            if not Path(file).suffix in VALID_SLIDE_EXTENSIONS: continue
+
+            slide_paths.append(file)
+
+
+    if debug_limit > 0: 
+        slide_paths = slide_paths[:debug_limit]
+        dry_run = True
+
+    if dry_run: logger.info ("Note, this is a dry run!!!")
+
+    logger.info(slide_paths)
+
+    write_adapter = IOAdapter(dry_run=dry_run).writer(store_url, bucket='pathology')
+
+    sp = SlideProcessor(writer=write_adapter, project=project_name)
+
+    with Client(host=os.environ['HOSTNAME'], n_workers=((num_cores + 1) // 2), threads_per_worker=2) as client:
+        logger.info ("Dashboard: "+ client.dashboard_link)
+        df = pd.DataFrame([x.result() for x in tqdm(as_completed(client.map(sp.run, slide_paths)), smoothing=0.01, total=len(slide_paths))]).set_index('slide_id')
+
+    df = df[df.index.notnull()]
+
+    rebase_schema_numeric(df)
+
+    df.loc[:, 'project_name'] = project_name
+    df.loc[:, 'comment'] = comment
+    df.loc[:, 'ingestion_date'] = datetime.today()
+
+    logger.info(df)
+
+    if dry_run:
+        return {}
+
+    output_table = os.path.join(output_dir, f"slide_ingest_{project_name}.parquet")
+    df.to_parquet(output_table)
+
+    properties = {
+        'slide_table': output_table
+    }
+
+    return properties
+
+
+class SlideProcessor:
+    def __init__(self, writer, project):
+        self.writer  = writer
+        self.project = project
+
+    def run(self, path):
+        """Extract openslide properties and write slide to storage location
+
+        Args:
+            path (string): path to slide image
+
+        Returns:
+            dict: slide metadata
+        """
+        try:
+            slide =  openslide.OpenSlide(path)
+        except:
+            print ("Couldn't process slide: ", path)
+            return {'slide_id':np.nan}
+        else:
+
+            input_slide_path  = Path(path)
+            
+            kv = dict(slide.properties)
+
+            kv['slide_id']   = str(input_slide_path.stem)
+            kv['slide_uuid'] = generate_uuid(path, ['WSI'])
+
+            write_kv = self.writer.write(path, f'{self.project}/slides')
```

Comment:
Recommend separating extracting openslide kv pairs / writing slide in 2 separate methods.